core/src/main/java/org/apache/struts2/components/template/FreemarkerTemplateEngine.java

```diff
@@ -144,15 +144,20 @@ public void close() throws IOException {
             }
         };
 
+        LOG.debug("Puts action on the top of ValueStack, just before the tag");
+        action = stack.pop();
         try {
             stack.push(templateContext.getTag());
+            stack.push(action);
```

Comment:
I realized right now that I think these two lines must be outside of the try after line 148.

Reply:
Could you elaborate a bit more?

Reply:
Yes sure thanks. I meant the `finally` block is here to undo those two pushes so consequently those two pushes must be done outside of the `try...finally`'s `try`. For example currently consider `stack.push(templateContext.getTag());` fails (it's unlikely but anyway just to describe) then the `finally` block pops three objects while we've only pushed two objects.

Reply:
Hm.. right, unlikely to happen but let's be safe, fixed!